### PR TITLE
Add Lelantus toggle to Firo wallets

### PR DESCRIPTION
**Pull request by sneurlax**
Scanning for Lelantus transactions may be disabled and is disabled by default.

TODO:
 - [x] Restore logic
 - [x] Refresh logic
 - [x] Desktop restore UI
 - [x] Mobile restore UI
 - [x] Desktop Lelantus settings UI
 - [x] Mobile Lelantus settings UI

Screenshot from 2024-05-29 19-51-21
Screenshot from 2024-05-29 17-30-21
Screenshot from 2024-05-29 19-38-22
Screenshot from 2024-05-29 19-38-16

### CODE REVIEW

File: lib/pages/add_wallet_views/restore_wallet_view/restore_options_view/restore_options_view.dart [hunk 1]
```diff
@@ -437,6 +441,32 @@ class _RestoreOptionsViewState extends ConsumerState<RestoreOptionsView> {
                     color: Colors.transparent,
                     child: Column(
                       children: [
+                        Row(
+                          children: [
+                            Checkbox(
```
Comment (julian-CStack) [inline lib/pages/add_wallet_views/restore_wallet_view/restore_options_view/restore_options_view.dart:446]: We have https://github.com/cypherstack/stack_wallet/blob/staging/lib/widgets/custom_buttons/checkbox_text_button.dart which might be easier to use here

File: lib/pages_desktop_specific/my_stack_view/wallet_view/sub_widgets/more_features/more_features_dialog.dart [hunk 2]
```diff
@@ -61,6 +69,13 @@ class _MoreFeaturesDialogState extends ConsumerState<MoreFeaturesDialog> {
       ),
     );
 
+    // Parse otherDataJsonString to get the enableLelantusScanning value.
+    if (wallet.info.otherDataJsonString != null) {
+      final otherDataJson = json.decode(wallet.info.otherDataJsonString!);
+      enableLelantusScanning =
+          otherDataJson["enableLelantusScanning"] as bool? ?? false;
```
Comment (julian-CStack) [inline lib/pages_desktop_specific/my_stack_view/wallet_view/sub_widgets/more_features/more_features_dialog.dart:76]: Use `WalletInfoKeys.enableLelantusScanning` in place of the string key. Even if the same now, a change to only one could introduce a hard to trace bug.

File: lib/pages_desktop_specific/my_stack_view/wallet_view/sub_widgets/more_features/more_features_dialog.dart [hunk 3]
```diff
@@ -53,6 +59,8 @@ class MoreFeaturesDialog extends ConsumerStatefulWidget {
 }
 
 class _MoreFeaturesDialogState extends ConsumerState<MoreFeaturesDialog> {
+  bool? enableLelantusScanning = false;
```
Comment (julian-CStack) [inline lib/pages_desktop_specific/my_stack_view/wallet_view/sub_widgets/more_features/more_features_dialog.dart:62]: This probably shouldn't be null. Could also be made private.

File: lib/pages_desktop_specific/my_stack_view/wallet_view/sub_widgets/more_features/more_features_dialog.dart [hunk 4]
```diff
@@ -136,6 +151,58 @@ class _MoreFeaturesDialogState extends ConsumerState<MoreFeaturesDialog> {
...
+                      width: 40,
+                      child: DraggableSwitchButton(
+                        isOn: enableLelantusScanning ?? false,
+                        onValueChanged: (newValue) {
+                          // Toggle enableLelantusScanning in wallet info.
+                          wallet.info.updateOtherData(newEntries: {
+                            WalletInfoKeys.enableLelantusScanning:
+                                !(enableLelantusScanning ?? false)
+                          }, isar: ref.read(mainDBProvider).isar).then((value) {
+                            // Should setState be used here?
```
Comment (julian-CStack) [inline lib/pages_desktop_specific/my_stack_view/wallet_view/sub_widgets/more_features/more_features_dialog.dart:179]: Yes probably. I would also use a simple bool value mutex to lock this function from being called again until it has completed.

File: lib/wallets/isar/models/wallet_info.dart [hunk 5]
```diff
@@ -507,4 +507,5 @@ abstract class WalletInfoKeys {
   static const String tezosDerivationPath = "tezosDerivationPathKey";
   static const String lelantusCoinIsarRescanRequired =
       "lelantusCoinIsarRescanRequired";
+  static const String enableLelantusScanning = "enableLelantusScanning";
```
Comment (julian-CStack) [inline lib/wallets/isar/models/wallet_info.dart:510]: Just for consistency I would add "Key" to the end of the string so its consistent with the others, not technically required though

File: lib/wallets/wallet/impl/firo_wallet.dart [hunk 6]
```diff
@@ -610,13 +610,25 @@ class FiroWallet<T extends ElectrumXCurrencyInterface> extends Bip39HDWallet<T>
           await mainDB.deleteWalletBlockchainData(walletId);
         }
 
+        // Parse otherDataJsonString to get the enableLelantusScanning value.
+        bool? enableLelantusScanning = false;
+        if (info.otherDataJsonString != null) {
+          final otherDataJson = json.decode(info.otherDataJsonString!);
+          enableLelantusScanning =
+              otherDataJson["enableLelantusScanning"] as bool? ?? false;
```
Comment (julian-CStack) [inline lib/wallets/wallet/impl/firo_wallet.dart:618]: use constant instead of string key here

**Review by julian-CStack — CHANGES_REQUESTED** [review]

File: lib/pages_desktop_specific/my_stack_view/wallet_view/sub_widgets/more_features/more_features_dialog.dart (hunk 2 above)
Comment (sneurlax) [reply lib/pages_desktop_specific/my_stack_view/wallet_view/sub_widgets/more_features/more_features_dialog.dart:76]: Also applied anywhere else I used "enableLelantusScanning" vs WalletInfoKeys.enableLelantusScanning

File: lib/pages_desktop_specific/my_stack_view/wallet_view/sub_widgets/more_features/more_features_dialog.dart (hunk 4 above)
Comment (sneurlax) [reply lib/pages_desktop_specific/my_stack_view/wallet_view/sub_widgets/more_features/more_features_dialog.dart]: Resolved by aad422d92e808384068273e7f8ef0627161d711c

File: lib/pages/add_wallet_views/restore_wallet_view/restore_options_view/restore_options_view.dart (hunk 1 above)
Comment (sneurlax) [reply lib/pages/add_wallet_views/restore_wallet_view/restore_options_view/restore_options_view.dart:446]: Would probably be more consistent, UI-wise, anyways.  One min

File: lib/pages/settings_views/wallet_settings_view/wallet_settings_wallet_settings/lelantus_settings_view.dart [hunk 7]
```diff
@@ -105,16 +107,25 @@ class _LelantusSettingsViewState extends ConsumerState<LelantusSettingsView> {
                     height: 20,
                     width: 40,
                     child: DraggableSwitchButton(
-                      isOn: enableLelantusScanning ?? false,
-                      onValueChanged: (newValue) {
+                      isOn: _enableLelantusScanning,
+                      onValueChanged: (newValue) async {
+                        if (_isUpdatingLelantusScanning) return;
+                        setState(() {
+                          _isUpdatingLelantusScanning = true;
```
Comment (julian-CStack) [inline lib/pages/settings_views/wallet_settings_view/wallet_settings_wallet_settings/lelantus_settings_view.dart:114]: setState not needed here unless the ui needs to be redrawn based on that value being changed

**Review by julian-CStack — APPROVED** [review]

**Pull request merged**